### PR TITLE
Fix defederate loop

### PR DESCRIPTION
**Pull request by lepsa**
https://wearezeta.atlassian.net/browse/WPB-3668

Fixing how errors are caught, mapped, and logged.

Removed redundant errors from `deleteFederationDomainRemoteUserFromLocalConversations`
Removing redundant constraints from several functions.
Removing commented code that isn't needed.
Updating `deleteFederationDomainOneOnOne` to have a recovering policy so it keeps retrying when there is an error returned over HTTP from Brig.

## Checklist

 - [:heavy_check_mark:] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [:heavy_check_mark:] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/galley/src/Galley/API/Internal.hs [hunk 1]
```diff
@@ -646,8 +630,13 @@ deleteFederationDomainLocalUserFromRemoteConversation dom = do
...
   for_ (Map.toList rCnvMap) $ \(cnv, lUsers) -> do
-    -- All errors, either exceptions or Either e, get thrown into IO
-    mapError @NoChanges (const (InternalErrorWithDescription "No Changes: Could not remove a local member from a remote conversation.")) $ do
+    let catchBaddies =
+          logAndIgnoreErrors @NoChanges
+            ( const "NoChanges: Could not remove a local member from a remote conversation."
+            -- `NoChanges` doesn't contain too many details, so no point in showing it here.
+            )
+            "Federation domain removal"
+    catchBaddies $ do
```
Comment (fisx) [inline services/galley/src/Galley/API/Internal.hs:639]: mostly to minimize whitespace changes, but i also like the name.  :)

File: services/galley/src/Galley/API/Internal.hs [hunk 2]
```diff
@@ -572,38 +570,25 @@ deleteFederationDomainRemoteUserFromLocalConversations dom = do
...
+          Sem (Error NoChanges ': ErrorS 'NotATeamMember ': r) () ->
+          Sem r ()
+        mapAllErrors msgText =
+          -- This can be thrown in `updateLocalConversationUserUnchecked @'ConversationDeleteTag`.
+          P.logAndIgnoreErrors @(Tagged 'NotATeamMember ()) (const "Not a team member") msgText
+            -- This can be thrown in `updateLocalConversationUserUnchecked @'ConversationRemoveMembersTag`
+            . P.logAndIgnoreErrors @NoChanges (const "No changes") msgText
+
+    mapAllErrors "Federation domain removal" $ do
+      getConversation cnvId
```
Comment (fisx) [inline services/galley/src/Galley/API/Internal.hs:584]: more straight-forward than extracting cnvId from lConv again.

**Comment by fisx** [comment]
oh, and i may have removed some comments that i didn't understand, @lepsa could you double-check?

**Comment by fisx** [comment]
another thing: is there a way to exponentially backoff rabbitMQ directly?  this would be the most robust way of making sure these crash loops are less aggressive, while still maintaining the same response times in almost all cases.

File: services/galley/src/Galley/API/Internal.hs [hunk 3]
```diff
@@ -653,25 +636,28 @@ deleteFederationDomainLocalUserFromRemoteConversation dom = do
...
-                }
-        -- These functions are used directly rather than as part of a larger conversation
-        -- delete function, as we don't have an originating user, and we can't send data
-        -- to the remote backend.
-        -- We don't need to check the conversation type here, as we can't tell the
-        -- remote federation server to delete the conversation. They will have to do a
-        -- similar processing run for removing the local domain from their federation list.
-        onConversationUpdated dom convUpdate
+    logAndIgnoreErrors @NoChanges
+      (const "No Changes: Could not remove a local member from a remote conversation.")
```
[the remark this answers is not in the document]
Comment (lepsa) [reply services/galley/src/Galley/API/Internal.hs]: NoChanges doesn't hold any information itself beyond the 0 arity data structure. We could use it in a `Show` context, but that would make the error message a bit less nice to read.

The same idea holds for NotATeamMember. Only 2 of the `GalleyError` constructors hold additional information, and we aren't catching either in this code.

File: services/galley/src/Galley/API/Internal.hs [hunk 4]
```diff
@@ -680,18 +666,12 @@ deleteFederationDomainLocalUserFromRemoteConversation dom = do
...
-  liftIO (deleteFederationRemoteGalley dom c)
-    >>= either
-      ( \e -> do
-          P.err $ Log.msg @String "Could not delete one-on-one messages in Brig" . Log.field "error" (show e)
-          -- Throw the error into IO to match the other functions and to prevent the
-          -- message from rabbit being ACKed.  TODO: should still be acked, or loop!
-          liftIO $ throwIO e
-      )
-      pure
+      -- This is the same policy as background-worker for retrying.
```
[the remark this answers is not in the document]
Comment (lepsa) [reply services/galley/src/Galley/API/Internal.hs:670]: Doing it here in addition to in background worker will help speed things up, specifically when Brig returns an error as we try to delete the connections. Before, this error would be thrown all the way up to servant where it would make Brig retry the galley call. This means galley would have to do all the DB work of looking for the remote domain. This isn't much of a cost, but it is a cost nevertheless. This is basically short-circuiting that retrying logic here to keep things running a bit faster. We still want the recovering logic in background-worker in case the network drops, or something else goes wrong in the code.

I copied the policy from background worker because it seemed reasonable and I didn't want to experiment with new values when I'm trying to cover a stricter subset of errors.
After 60 seconds the retry delay will be at a constant 60 seconds.

File: services/galley/src/Galley/API/Internal.hs (hunk 1 above)
Comment (lepsa) [reply services/galley/src/Galley/API/Internal.hs:639]: Both are nice reasons! The name is short and to the point.

**Comment by lepsa** [comment]
> another thing: is there a way to exponentially backoff rabbitMQ directly? this would be the most robust way of making sure these crash loops are less aggressive, while still maintaining the same response times in almost all cases.

We can back it off using dead letter queues and timeouts, but we will have to change some of the semantics around how we NACK messages we can't parse.

The basic idea for backing off messages is to have two queues, as explained here https://medium.com/@dotbox/delayed-requeuing-with-rabbitmq-dcbdf0026bf0

TL;DR, have 2 queues. A main one you pull from and a dead letter queue. The dead letter queue has a TTL timeout and has it's own dead lettering set to push the messages back to the main queue.

When a message is NACKed by the consumer, it is put into the dead letter queue. After the timeout Rabbit will then put it back into the main queue where it can be delivered again. Since rabbit supports multiple queues dead-lettering to the same queue, I'm sure you could do something with the message headers and message exchanges to get pseudo-exponential backoff by having a set of queues, i.e. a 10s timeout, a 20 second timeout, 40s, ...

**Comment by lepsa** [comment]
Changes look good to me

File: services/galley/src/Galley/API/Internal.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/API/Internal.hs]: sorry, i meant to remove that question.  all good!

**Review by fisx — APPROVED** [review]
